Add the concept of schema:dump to data migrations

tobyndockerill opened the pull request:
This commit introduces the concept of storing the current data migration version in `db/data_schema.rb`, and keeping this up to date as data migrations are run.

See https://github.com/ilyakatz/data-migrate/issues/65

ilyakatz commented on lib/data_migrate/data_migrator.rb:
```diff
@@ -81,4 +97,90 @@ def db_config
+
+  ##
+  # This class extends DatabaseTasks to override the schema_file method.
+  class DatabaseTasks
```
how about moving this into it's own file?

ilyakatz commented on lib/data_migrate/data_migrator.rb:
```diff
@@ -81,4 +97,90 @@ def db_config
+  ##
+  # Provides the capability to write the current data schema version to the data_schema file
+  # Based on ActiveRecord::SchemaDumper
+  class SchemaDumper
```
same here, separate file please just to keep things clean (i've been trying to split large files into smaller ones. as well as

ilyakatz commented on tasks/databases.rake:
```diff
@@ -223,6 +228,17 @@ namespace :db do
+
+  namespace :schema do
+    namespace :load do
+      desc 'Load both schema.rb and data_schema.rb file into the database'
```
For some reason I don't see this task

```
rake data:forward                 # Pushes the schema to the next version (specify steps w/ STEP=n)
rake data:migrate                 # Migrate data migrations (options: VERSION=x, VERBOSE=false)
rake data:migrate:down            # Runs the "down" for a given migration VERSION
rake data:migrate:redo            # Rollbacks the database one migration and re migrate up (options: STEP=x, VERSION=x)
rake data:migrate:status          # Display status of data migrations
rake data:migrate:up              # Runs the "up" for a given migration VERSION
rake data:rollback                # Rolls the schema back to the previous version (specify steps w/ STEP=n)
rake data:version                 # Retrieves the current schema version number for data migrations
rake db:forward:with_data         # Pushes the schema to the next version (specify steps w/ STEP=n)
rake db:migrate:down:with_data    # Runs the "down" for a given migration VERSION
rake db:migrate:redo:with_data    # Rollbacks the database one migration and re migrate up (options: STEP=x, VERSION=x)
rake db:migrate:status:with_data  # Display status of data and schema migrations
rake db:migrate:up:with_data      # Runs the "up" for a given migration VERSION
rake db:migrate:with_data         # Migrate the database data and schema (options: VERSION=x, VERBOSE=false)
rake db:rollback:with_data        # Rolls the schema back to the previous version (specify steps w/ STEP=n)
rake db:version:with_data         # Retrieves the current schema version numbers for data and schema migrations
```

ilyakatz commented:
@tobyndockerill sorry for the radio silence! I finally found some time to look at this and studied the code. I added a few commits. Things will go quicker from this point since I have some background

tobyndockerill commented:
Hi @ilyakatz, I've addressed your comments in this latest commit. I wasn't able to reproduce the issue you found when running `rake -T | grep data` however. I'm seeing the correct task in my list. Are you able to try again and let me know?

Cheers

ilyakatz commented:
Thanks @tobyndockerill, nice refactor!

I was testing this out and ran into an issue while running 

```
bundle exec rake db:create db:migrate          
```
It run migration successfully and then bails out because it's trying to do data schema dump without data_migration table. Do you want data schema dumper to run even if `data:migrate` hasn't run?

```
== 20171004211037 CreateIssueStates: migrating ================================
-- create_table(:issue_states)
   -> 0.0256s
== 20171004211037 CreateIssueStates: migrated (0.0257s) =======================

rake aborted!
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "data_migrations" does not exist
LINE 8:                WHERE a.attrelid = '"data_migrations"'::regcl...
                                          ^
:               SELECT a.attname, format_type(a.atttypid, a.atttypmod),
                     pg_get_expr(d.adbin, d.adrelid), a.attnotnull, a.atttypid, a.atttypmod,
                     c.collname, col_description(a.attrelid, a.attnum) AS comment
                FROM pg_attribute a
                LEFT JOIN pg_attrdef d ON a.attrelid = d.adrelid AND a.attnum = d.adnum
                LEFT JOIN pg_type t ON a.atttypid = t.oid
                LEFT JOIN pg_collation c ON a.attcollation = c.oid AND a.attcollation <> t.typcollation
               WHERE a.attrelid = '"data_migrations"'::regclass
                 AND a.attnum > 0 AND NOT a.attisdropped
               ORDER BY a.attnum
/Users/katz/ws/data-migrate/lib/data_migrate/schema_dumper.rb:33:in `initialize'
/Users/katz/ws/data-migrate/lib/data_migrate/schema_dumper.rb:12:in `new'
/Users/katz/ws/data-migrate/lib/data_migrate/schema_dumper.rb:12:in `dump'
/Users/katz/ws/data-migrate/tasks/databases.rake:349:in `block (3 levels) in <top (required)>'
/Users/katz/ws/data-migrate/tasks/databases.rake:348:in `open'
/Users/katz/ws/data-migrate/tasks/databases.rake:348:in `block (2 levels) in <top (required)>'
/Users/katz/ws/data-migrate/tasks/databases.rake:3:in `block (2 levels) in <top (required)>'
/Users/katz/.rbenv/versions/2.4.2/bin/bundle:23:in `load'
/Users/katz/.rbenv/versions/2.4.2/bin/bundle:23:in `<main>'
PG::UndefinedTable: ERROR:  relation "data_migrations" does not exist
LINE 8:                WHERE a.attrelid = '"data_migrations"'::regcl...
                                          ^
/Users/katz/ws/data-migrate/lib/data_migrate/schema_dumper.rb:33:in `initialize'
/Users/katz/ws/data-migrate/lib/data_migrate/schema_dumper.rb:12:in `new'
/Users/katz/ws/data-migrate/lib/data_migrate/schema_dumper.rb:12:in `dump'
/Users/katz/ws/data-migrate/tasks/databases.rake:349:in `block (3 levels) in <top (required)>'
/Users/katz/ws/data-migrate/tasks/databases.rake:348:in `open'
/Users/katz/ws/data-migrate/tasks/databases.rake:348:in `block (2 levels) in <top (required)>'
/Users/katz/ws/data-migrate/tasks/databases.rake:3:in `block (2 levels) in <top (required)>'
/Users/katz/.rbenv/versions/2.4.2/bin/bundle:23:in `load'
/Users/katz/.rbenv/versions/2.4.2/bin/bundle:23:in `<main>'
Tasks: TOP => data:dump
(See full trace by running task with --trace)
```

tobyndockerill commented:
That's a good point, I think there's an edge case here that I didn't quite think of.

After migrating the database Rails dumps the contents of the newly-migrated database to `schema.rb` (to ensure that everything's in sync).

As part of this PR I added the following code to ensure that when the database schema is dumped, we also dump the data schema:
```ruby
Rake::Task["db:_dump"].enhance do
  Rake::Task["data:dump"].invoke
end
``` 
This works as long as a data-specific migration task has been called beforehand, as many of these tasks call `assure_data_schema_table` beforehand.

In the case you've identified, none of these tasks are run before the `data:dump` task is called.

Perhaps a call to `assure_data_schema_table` should be added to the `data:dump` task. What are your thoughts?

ilyakatz commented:
Hm, I'm wondering if db:migrate should even know anything about data:migrate. For example, let's say you added `data-migrate` gem to Gemfile but didn't actually do any data migrations yet. This will introduce dependency which is not necessary. What if we only do dumps on data:migrate related rake tasks?

tobyndockerill commented:
That's a good call, I'll remove the code as above and add an explicit call to `data:dump` in each `data:migrate`-related task.

ilyakatz commented:
Thanks @tobyndockerill, you've been great on this PR!

I just merged and released 3.3.0 with your changes

https://github.com/ilyakatz/data-migrate/blob/master/Changelog.md#330

tobyndockerill commented:
Cheers @ilyakatz!
